### DISCUSSION TITLE
2D NFFT and get 2D array back?

### DISCUSSION THREAD

**Discussion by roflmaostc** [JuliaMath/NFFT.jl#64] [Q&A] [answered]
Hi!

is it possible to generate the NFFT of an array and get an array back again?
As far as I understand the examples, you basically pass a linear list of frequencies to the NFFT routine and get a vector of Fourier amplitudes back.

But similar to FFT I would like to have a grid structure, so a 2D array back. 

Best,

Felix

**Comment by tknopp** [comment, marked as answer]
how should that work? This can only work if the nodes have some grid structure.

**Comment by JeffFessler** [comment]
You could use `reshape` of course, or you could automate this by using a `LinearMapAO` type https://github.com/JeffFessler/LinearMapsAA.jl
There is a somewhat related example here for radial sampling, which is a typical case where the output has some structure:
https://juliaimagerecon.github.io/Examples/generated/mri/3-2d-t/

**Comment by tknopp** [comment]
yes, my point is that this only works if the nodes `x` represent a grid. And this means it should be handled via a reshape. That is also important to remain typestable. The outputdim needs to be inferable from all the input data.

**Comment by roflmaostc** [comment]
Yeah, my output has structure (some scaling basically).

I understand your points, makes sense.

Thanks!

**Comment by RainerHeintzmann** [comment]
... this is now simplified by using the `nfft_nd` function in the FourierTools packagem, currently branch:
https://github.com/bionanoimaging/FourierTools.jl/tree/resample_var

By the way, it is not always only a reshape, if some coordinates refer to out-of-bounds positions (adjunct operator). This is handled via considering only a subset of points in the source array (adjunct operator). The reshape on the input becomes rather indexing via a binary array along with a list of only in-bound positions.
